src/resourceClient/baseResourceClient.ts

```diff
@@ -1,240 +1,287 @@
-    return new Promise(resolve => setTimeout(resolve, delay));
-  }
-}
-
-const datePattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
-
-function parseJSONWithDateHandling(json: string) {
-  return JSON.parse(json, (_key: any, value: any) => {
-    if (typeof value !== 'string') {
-      return value;
-    }
-
-    const calendarDate = CalendarDate.parse(value);
-
-    if (calendarDate) {
-      return calendarDate;
-    }
-
-    if (RegExp(datePattern).exec(value)) {
-      return new Date(value);
-    }
-
-    return value;
-  });
-}
+import axios, {
+  AxiosError,
+  AxiosInstance,
+  AxiosPromise,
+  AxiosRequestConfig,
+  AxiosResponse,
+  Method
+} from 'axios';
+import * as https from 'https';
+import axiosBetterStacktrace from 'axios-better-stacktrace';
+import { ClientConfiguration } from '../clientConfiguration';
+import { CalendarDate } from '../calendarDate';
+import ApiUrlBuilder from './apiUrlBuilder/apiUrlBuilder';
+import { ClientRequest, IncomingMessage } from 'http';
+import {
+  BadRequestError,
+  ForbiddenError,
+  NotFoundError,
+  RatelimitError,
+  RequestRetriesExceededError,
+  UnauthorizedError
+} from './errors';
+
+export class BaseResourceClient {
+  protected client: AxiosInstance;
+  baseURL: string;
+
+  constructor(
+    public urlBuilder: ApiUrlBuilder,
+    public config: ClientConfiguration
+  ) {
+    this.baseURL = config.http?.baseURL ?? 'https://api.music.apple.com';
+    this.client = this.create({ resInterceptor: true });
+  }
+
+  // create axios client, set interceptors, handle errors & auth
+  private create(options: { resInterceptor?: boolean }): AxiosInstance {
+    const config: AxiosRequestConfig = {
+      baseURL: this.baseURL,
+      proxy: this.config.http?.proxy,
+      // https://github.com/axios/axios/blob/v0.20.0-0/lib/defaults.js#L57-L65
+      transformResponse: [
+        (data) => {
+          /*eslint no-param-reassign:0*/
+          if (typeof data === 'string') {
+            try {
+              data = parseJSONWithDateHandling(data);
+            } catch (e) {
+              /* Ignore */
+            }
+          }
+          return data;
+        }
+      ],
+      validateStatus: () => true // Handle errors by ourselves
+    };
+
+    if (this.config.http?.localAddress) {
+      config.transport = {
+        ...https,
+        request: (
+          options: https.RequestOptions,
+          callback: (res: IncomingMessage) => void
+        ): ClientRequest =>
+          https.request(
+            {
+              ...options,
+              localAddress: this.config.http!.localAddress,
+              family: this.config.http!.localAddress?.includes(':') ? 6 : 4
+            },
+            callback
+          )
+      };
+    }
+
+    const client = axios.create(config);
+    axiosBetterStacktrace(client);
+
+    client.interceptors.request.use((config) => {
+      config.headers.Authorization = `Bearer ${this.config.developerToken}`;
```

Comment:
previous version I liked more

Reply:
The previous one was erroring with the following, so I opted for a quick switch. In the new types it can either be a raw object or a map, and typescript wasn't liking the old code.

```
Type '{ Accept?: AxiosHeaderValue; "Content-Length"?: AxiosHeaderValue; 'User-Agent': AxiosHeaderValue; "Content-Encoding"?: AxiosHeaderValue; Authorization: AxiosHeaderValue; 'Content-Type'?: ContentType; 'media-user-token': string; origin: string; }' is not assignable to type 'AxiosRequestHeaders'.
  Type '{ Accept?: AxiosHeaderValue; "Content-Length"?: AxiosHeaderValue; 'User-Agent': AxiosHeaderValue; "Content-Encoding"?: AxiosHeaderValue; Authorization: AxiosHeaderValue; 'Content-Type'?: ContentType; 'media-user-token': string; origin: string; }' is missing the following properties from type 'AxiosHeaders': set, get, has, delete, and 23 more.
```